src/Core/src/Response.php

```diff
@@ -85,24 +86,74 @@ public function resolve(?float $timeout = null): bool
                 }
             }
 
-            $statusCode = $this->httpResponse->getStatusCode();
+            return $this->handleStatus();
         } catch (TransportExceptionInterface $e) {
             throw $this->resolveResult = new NetworkException('Could not contact remote server.', 0, $e);
         }
+    }
 
-        if (500 <= $statusCode) {
-            throw $this->resolveResult = new ServerException($this->httpResponse);
-        }
+    /**
+     * Make sure all provided requests are executed.
+     *
+     * @param self[]     $responses
+     * @param float|null $timeout   Duration in seconds before aborting. When null wait
+     *                              until the end of execution. Using 0 means non-blocking
+     *
+     * @return iterable<self>
+     *
+     * @throws NetworkException
+     * @throws HttpException
+     */
+    final public static function multiplex(iterable $responses, float $timeout = null): iterable
+    {
+        /** @var self[] $responseMap */
+        $responseMap = [];
+        $httpResponses = [];
+        $httpClient = null;
+        foreach ($responses as $response) {
+            if (null !== $response->resolveResult) {
+                yield $response;
+
+                continue;
+            }
 
-        if (400 <= $statusCode) {
-            throw $this->resolveResult = new ClientException($this->httpResponse);
+            if (null === $httpClient) {
+                $httpClient = $response->httpClient;
+            }
+            $httpResponses[] = $response->httpResponse;
+            $responseMap[\spl_object_id($response->httpResponse)] = $response;
         }
 
-        if (300 <= $statusCode) {
-            throw $this->resolveResult = new RedirectionException($this->httpResponse);
+        // response is empty
+        if (null === $httpClient) {
+            return;
         }
 
-        return $this->resolveResult = true;
+        try {
+            foreach ($httpClient->stream($httpResponses, $timeout) as $httpResponse => $chunk) {
+                if ($chunk->isTimeout()) {
+                    continue;
+                }
+                if ($chunk->isFirst()) {
+                    $response = $responseMap[\spl_object_id($httpResponse)] ?? null;
+
+                    if (null !== $response) {
+                        try {
+                            $response->handleStatus();
+                        } catch (Exception $e) {
+                        }
+                        yield $response;
+                    }
+                    unset($responseMap[\spl_object_id($httpResponse)]);
+                    if (empty($responseMap)) {
+                        // early exit if all statusCode are known. We don't have to wait for all responses bodies
+                        return;
+                    }
+                }
+            }
+        } catch (TransportExceptionInterface $e) {
+            throw new NetworkException('Could not contact remote server.', 0, $e);
```

Comment:
Hm.. When will this be thrown?

Reply:
This catch should be inside the loop. 

One request may throw a TransportException but that should not effect all of them, right?

Reply:
it throw by iterrating on `$httpClient->stream()`

> One request may throw a TransportException

yes you're right. Should I play with `next`/`current` to manually move the cursor?

Reply:
hmm, we can't do it.
nb: in fact nothing can throw an exception inside the loop (we just manipulat DTO objects)

The exception can be thrown between 2 yield of HttpClient: when moving the cursor from a response to another (internally calling `iterator::next`).
If such exception occurred we can not ignore this failed response and witch to another, given the exception is thrown while moving the cursor.

Reply:
Just figured out that `$chunk->timeout()` (and other methods of chunk) will throw the exception.
It's possible to continue the loop. I've fixed it in the last commit